doc/sphinx/proof-engine/ltac2.rst

```diff
@@ -659,168 +748,459 @@ insert in a concise way an Ltac2 variable of type :n:`constr` into a Coq term.
+   expression.  Matching is first-order except on variables of the form :n:`@?@ident`
+   that occur in the head position of an application. For these variables,
+   matching is second-order and returns a functional term.
+
+   .. todo the `@?ident` form is in dangling_pattern_extension_rule, not included in the doc yet
+      maybe belongs with "Applications"
+
+   `lazy_match!`
+      Causes the match to commit to the first matching branch
+      rather than trying a new match if :n:`@ltac2_expr` fails.
+      :ref:`Example<ltac2_match_vs_lazymatch_ex>`.
+
+   `match!`
+      If :n:`@ltac2_expr` fails, continue matching with the next branch.
+      Failures in subsequent tactics (after the `match!`) will not cause selection
+      of a new branch.  Examples :ref:`here<ltac2_match_vs_lazymatch_ex>` and
+      :ref:`here<ltac2_match_vs_multimatch_ex>`.
+
+   `multi_match!`
+      If :n:`@ltac2_expr` fails, continue matching with the next branch.
+      When a :n:`@ltac2_expr` succeeds for a branch, subsequent failures
+      (after the `multi_match!`) causing consumption of all the successes
+      of :n:`@ltac2_expr` trigger selection of a new matching branch.
+      :ref:`Example<ltac2_match_vs_multimatch_ex>`.
+
+   :n:`@cpattern`
+      The syntax of :token:`cpattern` is
+      the same as that of :token:`term`\s, but it can contain pattern matching
+      metavariables in the form :n:`?@ident`.  :g:`_` can be used to match
+      irrelevant terms.
+
+      .. todo Example is broken :ref:`Example<ltac2_match_with_holes_ex>`.
+
+      .. todo Didn't understand the following 2 paragraphs well enough to revise
+         see https://github.com/coq/coq/pull/12103#discussion_r436297754 for a
+         possible example
+
+      When a metavariable in the form :n:`?id` occurs under binders,
+      say :n:`x__1, …, x__n` and the expression matches, the
+      metavariable is instantiated by a term which can then be used in any
+      context which also binds the variables :n:`x__1, …, x__n` with
+      same types. This provides with a primitive form of matching under
+      context which does not require manipulating a functional term.
+
+      There is also a special notation for second-order pattern matching: in an
+      applicative pattern of the form :n:`@?@ident @ident__1 … @ident__n`,
+      the variable :token:`ident` matches any complex expression with (possible)
+      dependencies in the variables :n:`@ident__i` and returns a functional term
+      of the form :n:`fun @ident__1 … @ident__n => @term`.
+
+   .. _match_term_context:
+
+   :n:`context {? @ident } [ @cpattern ]`
+      Matches any term with a subterm matching :token:`cpattern`. If there is a match
+      and :n:`@ident` is present, it is assigned the "matched
+      context", i.e. the initial term where the matched subterm is replaced by a
+      hole.  Note that `context`
+      (with very similar syntax) appearing after the `=>` is the :tacn:`context` tactic.
```

Comment:
There is no `context` tactic in Ltac2.
```suggestion
      hole.  This matched context has type `Pattern.context` and the hole
      can be filled with the function `Pattern.instantiate`.
```

Reply:
Your quotes are fine.  I said:

> This hole in the matched context can be filled with the expression
      :n:`Pattern.instantiate @ident @cpattern`.